File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> allStrings = new HashSet<>();
        int res = 0;
        int j = 0;
        int i = 0;
        while (i < s.length()) {
            char curr = s.charAt(i);
            if (!allStrings.contains(curr)) {
                allStrings.add(curr);
                i++;
                res = Math.max(res, allStrings.size());
            } else {
                allStrings.remove(s.charAt(j));
                j++;
            }
        }
        return res;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        i = 0
        charSet = set()

        for j in range(len(s)):
            while s[j] in charSet:
                charSet.remove(s[i])
                i += 1
            
            charSet.add(s[j])
            res = max(res, j - i + 1)            
        return res